### ISSUE TITLE
RavenDB 6.0.104 Select ID not working correctly when mapping to different name

### ISSUE THREAD

**Issue by xinix00** [ravendb/ravendb#18751]
Following output of query differs from 5.x to 6.0:
```
from result in session.Query<Index.Mapping, Index>()
select {
    Id = result.PurchaseId
    DistributedId = result.Id
}
```
Will work in previous versions, and will give the Id to the DistributedId. In new versions, DistributedId is NULL, however, when changing it to:
```
from result in session.Query<Index.Mapping, Index>()
select {
    Id = result.Id
    PurchaseId = result.PurchaseId
}
```
The Id will be filled.

RavenDB will translate it in both cases to:
```
declare function output(result, ...) {
    return { Id: result.PurchaseId, DistributedId: result.Id };
}
```
DistributedId = NULL (in Raven 6.0.104)
```
declare function output(result, ...) {
    return { Id: result.Id, PurchaseId: result.PurchaseId };
}
```
Id = filled

Though the output differs, yes I know, it doesn't seem to make sense.

**Comment by arekpalinski** [comment]
Hi Derek,

Can you please provide an index definition?

**Comment by xinix00** [comment]
```
AddMap<DistributedPurchase>(results => select new Mapping
{
    Id = result.Id
 })
 ```
 
 And I'm sorry, this was incomplete, I also do .As<ENTITY>(), so the above query should be replaced with:
```
 from result in session.Query<Index.Mapping, Index>().As<DistributedPurchase>()
 ```

**Comment by ayende** [comment]
We would need to see the full index, class definition and query code.
Can you try to create a failing unit test?

As a guess, maybe this is related to storing fields?

**Comment by xinix00** [comment]
You can see it in action in the studio RQL:
```
declare function output(result) {
    return { Id : result.Id };
} from index DistributedAll as result select { Id : result.Id }
```
Which was in C#:
```
from a in session
    .Query<DistributedAll.Mapping, DistributedAll>()
    .As<Distributed>()
select new { Id = result.Id } 
```
Id will now be mapped to the @id , though the "real" id is null. This is why DistributedId = result.Id will result in null, since it doesn't fall back to @id.

In RavenDB 5.4.x the same query will be translated to:
```
Id : id(result)
```

Which will work. Same input, different output. Or is yours translating it to id(result)? P.s. I'm projecting the document, not the index. If I check, I also see that RESULT is the document (without the ID, which is in METADATA), I think that's why earlier versions translated it to id(result).

**Comment by xinix00** [comment]
P.s. a real example could be as simple as this:
```

            var distributed2 = 
                from a in _session.Query<DistributedAll.Mapping, DistributedAll>().As<DistributedPurchase>()
                let articleSort = RavenQuery.Load<ArticleSort>(a.ArticleSortId)
                select new {
                    DistributedId = a.Id
                };

```
                
Which turns into:
```
from index 'DistributedAll' as a load a.ArticleSortId as articleSort select { DistributedId : a.Id }
```

I think the old behavior was better (translating it to the ID (so id(result)). This because I'm using the .As<>(), so I'm projecting the full document and transforming it. This object HAS the Id inside it, however, RavenDB is returning NULL at this very moment.

I hope you do get the problem now?

**Comment by arekpalinski** [comment]
@xinix00 I have reproduced the issue and reported it on our bugtracker: 
https://issues.hibernatingrhinos.com/issue/RavenDB-22556/JS-projection-query-uses-x.Id-on-6.0-while-it-used-idx-on-5.4

**Comment by Lwiel** [comment]
From version 6.0 we treat `x.Id` as identity property (`id(x)`) only in two cases (https://github.com/ravendb/ravendb/pull/16825):
 - It's a property accessed on a loaded document
 - It's a property accessed on a query root document
 
The reason for this is to correctly handle queries that access nested property named `Id`, e.g. `x.Something.Id`, which in this case should be treated as a normal property, not an identity.
 
In your case the query root type is `DistributedAll.Mapping`, defined in `session.Query` method. 
Method `As` doesn't change the query root type, it's suitable only for projecting the results. 

If you change your query to the following, it should work correctly. 
```csharp
var distributed2 =
                    from a in session.Query<DistributedPurchase, DistributedAll>()
                    let articleSort = RavenQuery.Load<ArticleSort>(a.ArticleSortId)
                    select new
                    {
                        DistributedId = a.Id
                    };
```

If you'd like to filter your results by some indexed field, you can do the following. Please note that it makes the query more complicated, and a much better solution would be to adjust your data model to queries that you're going to use.

```csharp
var distributed2 =
                    from a in session.Query<DistributedPurchase, DistributedAll>().As<DistributedAll.Mapping>().Where(x => x.Id == "DistributedPurchases/1").As<DistributedPurchase>()
                    let articleSort = RavenQuery.Load<ArticleSort>(a.ArticleSortId)
                    select new
                    {
                        DistributedId = a.Id
                    };
```

**Comment by xinix00** [comment]
```
var distributed2 =
                    from a in session.Query<DistributedPurchase, DistributedAll>().As<DistributedAll.Mapping>().Where(x => x.Id == "DistributedPurchases/1").As<DistributedPurchase>()
                    let articleSort = RavenQuery.Load<ArticleSort>(a.ArticleSortId)
                    select new
                    {
                        DistributedId = a.Id
                    };
```

But this is what I did, and also what arek has reproduced. I don't see the difference with what I reported with:
```
            var distributed2 = 
                from a in _session.Query<DistributedAll.Mapping, DistributedAll>()
                .As<DistributedPurchase>()
                select new {
                    DistributedId = a.Id
                };
```

**Comment by Lwiel** [comment]
The difference is the type passed to `session.Query<Type1, Type2>()`. In your latest comment the first query has `DistributedPurchase` as `Type1`, the second query has `DistributedAll.Mapping`.

It's the only place where you can define what will be the return type of the query. `As<Type>()` method was made only for projecting the end results, so if you use it in the middle of the query, Raven still operates on the type provided in `session.Query<Type1, Type2>()` method. 

The query that doesn't work has `DistributedAll.Mapping` defined as `Type1`, but you immediately use the `As<DistributedPurchase>()` method. Because of that your `a.ArticleSortId` comes from the document of type `DistributedPurchase`, but Raven expects it to come from `DistributedAll.Mapping` type. That's why it's not treated as an identity property.

**Comment by xinix00** [comment]
Haha thanks for pointing that out. I was used to setting the mapping in the query, I now see what you mean, since we normally are expecting the document as result (not the mapping except for map reduce) it seems better practice for me to always put the document type as TYPE1 for the query.

Did not noticed this was what you were trying to show. Anyhow thanks!

**Comment by xinix00** [comment]
@Lwiel, you also might want to change the documentation.

What you prefer is use .Query<DOCUMENT, INDEX>().As<MAPPING>()

But on multiple places in your documentation (almost all), you are using:
.Query<MAPPING, INDEX>().As<DOCUMENT>()

https://ravendb.net/docs/article-page/6.0/csharp/indexes/querying/suggestions
https://ravendb.net/docs/article-page/6.0/Csharp/indexes/querying/projections
https://ravendb.net/docs/article-page/6.0/Csharp/indexes/querying/sorting
https://ravendb.net/docs/article-page/6.0/Csharp/indexes/querying/query-index << Here you say explicitly to use the MAPPING, INDEX for querying (with filter) an index

And I can find much more places where you recommend the wrong <MAPPING, INDEX>. Maybe also fix your documentation?

**Comment by ayende** [comment]
See here: 
https://ravendb.net/learn/inside-ravendb-book/reader/4.0/12-working-with-indexes#working-with-complex-indexes-using-strongly-typed-code

There are the following models to consider:

- The documents to be indexed.
- The index entry that was outputted from the index.
- The actual queryable fields in the index.
- The result of the query.

**Comment by Lwiel** [comment]
In all of these examples projection (whether using `As<Type>()`, `ProjectInto<Type>()` or `OfType<Type>` method) is always done *after* the actual work (loading documents, searching, filtering, etc.) is done. Because the work is done only on indexed fields, using index entry class is enough. In such case the projection is used to return actual documents (including fields that are not indexed).

Take a look at [one of the examples you linked](https://ravendb.net/docs/article-page/6.0/Csharp/indexes/querying/query-index#query-an-index-by-query). In both queries we want to return documents of `Employee` type. 

The first query uses `Employee` type in `.Query<Employee, Employees_ByName>` method because there's no filtering. In this case there's no reason to project anything. 

The second query uses `Employees_ByName.IndexEntry` type because of filtering by `LastName` field (which is indexed). In this case it's necessary to project the end results into `Employee` type (here done by `.OfType<Employee>` method).

**Comment by xinix00** [comment]
I do get that, the only thing is that:
```
.Query<Employees_ByName.IndexEntry, Employees_ByName>()
.Where(x => x.LastName == "Rahien")
.As<Employee>()
.Select(a => new {
    FirstName = a.FirstName
    LastName = a.LastName
    EmployeeId = a.Id
});
```

Doesn't work anymore, while it did on version 5.4. The EmployeeId in this example will be null on 6.0.

You now have to do:
```
.Query<Employee, Employees_ByName>()
.As<Employees_ByName.IndexEntry>()
.Where(x => x.LastName == "Rahien")
.As<Employee>()
.Select(a => new {
    FirstName = a.FirstName
    LastName = a.LastName
    EmployeeId = a.Id
});
```

To get the right result. For me this isn't clear when you look at the documentation.

**Comment by Lwiel** [comment]
We decided that we'll add a dedicated method to allow users to explicitly project document id, here's a ticket for this -https://issues.hibernatingrhinos.com/issue/RavenDB-22624/Add-RavenQuery.Id-method-to-allow-explicit-identity-projection

**Comment by Lwiel** [comment]
Merged pull request with dedicated method (`RavenQuery.Id`) implementation - https://github.com/ravendb/ravendb/pull/19398
It's available in versions 6.2.2+
